.github/workflows/build-and-test.yml

```diff
@@ -5,115 +5,185 @@ on:
     branches: [main]
   pull_request:
     branches: [main]
+
+concurrency:
+  # Group concurrency on workflow, then:
+  # - Is merge run? Group on branch name (`refs/heads/main`)
+  # - Is pull request? Group on pull request branch name, for example `feat/add-awesome-feature`
+  group: >-
+    ${{ github.workflow }}-${{
+      github.event_name == 'push'
+        && github.ref
+        || github.head_ref
+    }}
+  # Run merge workflows in sequence to prevent parallel deployments and releases
+  # Cancel stale pull request runs in progress for the same branch
+  cancel-in-progress: ${{ github.event_name != 'push' }}
+
 env:
-  NX_CLOUD_DISTRIBUTED_EXECUTION: true
-  NX_DISTRIBUTED_TASK_EXECUTION: true
+  NODE_OPTIONS: --max-old-space-size=6144
   NX_VERBOSE_LOGGING: false
+  docs-name: docs
+  docs-path: dist/apps/docs
 
 jobs:
-  agents:
+  dte_agents:
+    name: DTE Agents
     runs-on: ubuntu-latest
-
-    name: Agent 1
     timeout-minutes: 20
 
+    # The Free GitHub plan has a limit of 20 concurrent jobs on Ubuntu images
+    # Reference: https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration
+    # If we need to, we can optimize for 2 simultaneous workflow runs:
+    # 2 x 1 main job   =  2 concurrent jobs
+    # 2 x 9 agent jobs = 18 concurrent jobs
+    # Total            = 20 concurrent jobs
+    #
+    # However, we don't have many projects or targets in this workspace, so we
+    # lower the number of agents to reduce spent compute time.
     strategy:
       matrix:
-        agent: [1, 2, 3]
+        agent: [1, 2, 3, 4]
 
     steps:
-      - uses: actions/checkout@v2
-      - uses: ./.github/actions/setup
+      - name: Checkout all commits
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Set up dependencies
+        uses: ./.github/actions/setup
 
-      - name: Start Nx Agent ${{ matrix.agent }}
+      - name: Start Nx Cloud DTE Agent ${{ matrix.agent }}
         run: yarn nx-cloud start-agent
 
-  build:
+  # We're using Nx Cloud for Distributed Task Execution
+  # Reference: https://nx.dev/using-nx/dte
+  #
+  # The coordinator outputs the combination of task outputs from the agents,
+  # both terminal and file outputs
+  dte_coordinator:
+    name: DTE Coordinator
     runs-on: ubuntu-latest
+    env:
+      NX_CLOUD_DISTRIBUTED_EXECUTION: true
+      NX_DISTRIBUTED_TASK_EXECUTION: true
 
     steps:
-      - uses: actions/checkout@v2
-      - uses: ./.github/actions/setup
-
-      - run: yarn nx-cloud start-ci-run
-
+      - name: Checkout all commits
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Set up dependencies
+        uses: ./.github/actions/setup
+
+      - name: Start Nx Cloud DTE
+        run: yarn nx-cloud start-ci-run --stop-agents-after="e2e" --agent-count=4
```

Comment:
why do we need to specify --stop-agents here? Isn't it possible we run into a race condition? Let's say e2e comes from cache and finishes instantly, but some build is still missing. We anyway stop the agents manually, right?

Reply:
also, do we need to specify the --agent-count here? never seen this before. can we have this as a variable at least?

Reply:
> why do we need to specify --stop-agents here? Isn't it possible we run into a race condition? Let's say e2e comes from cache and finishes instantly, but some build is still missing. We anyway stop the agents manually, right?

We don't need `--stop-agents-after` because we are always calling `nx-cloud stop-all-agents` in a job step. I will remove this.

Reply:
> also, do we need to specify the --agent-count here? never seen this before. can we have this as a variable at least?

In the workflow generated by `@nrwl/workspace:ci-workflow`, this is added. I'm think I read about it in the Nx docs but seems that it has since been removed. It's equivalent to setting the [`NX_CLOUD_DISTRIBUTED_EXECUTION_AGENT_COUNT`](https://nx.dev/nx-cloud/reference/env-vars#nx_cloud_distributed_execution_agent_count) environment variable and can prevent timing and queuing issues.